File: Algorithms/LinkedList/singlyList.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

class mylist
{
private:
    struct node
    {
        int val;
        node *next;
        node(int val) : val(val), next(nullptr) {}
    };
    int size;
    node *head;
    node *tail;

public:
    // constructor
    mylist()
    {
        size = 0;
        head = nullptr;
        tail = nullptr;
    }

    // push front
    void push_front(int val)
    {
        node *newNode = new node(val);
        newNode->next = head;
        head = newNode;
        tail = head;
        size++;
    }
    // push back
    void push_back(int val)
    {
        // base case when list is empty
        if (size == 0)
        {
            push_front(val);

            return;
        }

        node *newNode = new node(val);
        node *cur = head;
        while (cur->next != nullptr)
        {
            cur = cur->next;
        }
        cur->next = newNode;
        size++;
    }

    void push_backk(int val){
        if(size == 0) {
            push_front(val);
            return;
        }
        tail->next = new node(val);
        tail = tail->next;
    }

    void insert(int index, int val)
    {
        // base case:
        // same if index == 0 then use push_front
        if (index < 0 || index >= size)
            return;
        
        if (index == 0)
        {
            push_front(val);
            return;
        }
        node *cur = head;
        for (int i = 0; i < (index - 1); i++)
        {
            cur = cur->next;
        }
        node *newNode = new node(val);
        newNode->next = cur->next;
        cur->next = newNode;
        size++;
    }

    void print()
    {
        // print whole list
        node *cur = head;
        while (cur != nullptr)
        {
            cout << cur->val << ' ';
            cur = cur->next;
        }
        cout << '\n';
    }

    void erase(int index)
    {
        // if index == 0 make 2nd element as head head = head->next
        if (index < 0 || index >= size)
            return;
        node *todel;
        if (index == 0)
        {
            todel = head;
            head = head->next;
        }
        else
        {
            node *cur = head;
            for (int i = 0; i < (index - 1); i++)
            {
                cur = cur->next;
            }
            todel = cur->next;
            cur->next = cur->next->next;
        }
        delete todel;
        size--;
    }
};

signed main()
{
    mylist A;
    A.push_front(34);
    A.push_front(23);
    A.push_front(3453);
    A.push_front(56);
    A.push_back(42);
    A.insert(2, 123);
    A.print();
    A.erase(2);
    A.erase(1);
    A.print();
    return 0;
}

File: Algorithms/LinkedList/singlyList.py
class MyList:
    class Node:
        def __init__(self, val):
            self.val = val
            self.next = None
    
    def __init__(self):
        self.size = 0
        self.head = None
    
    def push_front(self, val):
        new_node = self.Node(val)
        new_node.next = self.head
        self.head = new_node
        self.size += 1
        
    def push_back(self, val):
        if self.size == 0:
            self.push_front(val)
            return
        new_node = self.Node(val)
        cur = self.head
        while cur.next is not None:
            cur = cur.next
        cur.next = new_node
        self.size += 1
        
    def insert(self, index, val):
        if index < 0 or index >= self.size:
            return
        if index == 0:
            self.push_front(val)
            return
        cur = self.head
        for i in range(index - 1):
            cur = cur.next
        new_node = self.Node(val)
        new_node.next = cur.next
        cur.next = new_node
        self.size += 1
        
    def print_list(self):
        cur = self.head
        while cur is not None:
            print(cur.val, end=' ')
            cur = cur.next
        print()
        
    def erase(self, index):
        if index < 0 or index >= self.size:
            return
        to_del = None
        if index == 0:
            to_del = self.head
            self.head = self.head.next
        else:
            cur = self.head
            for i in range(index - 1):
                cur = cur.next
            to_del = cur.next
            cur.next = cur.next.next
        self.size -= 1
        del to_del
        
# Example Usage:
A = MyList()
A.push_front(34)
A.push_front(23)
A.push_front(3453)
A.push_front(56)
A.push_back(42)
A.insert(2, 123)
A.print_list()
A.erase(2)
A.erase(1)
A.print_list()
